src/python/pants/backend/jvm/tasks/coverage/jacoco.py

```diff
@@ -61,13 +100,69 @@ def classpath_prepend(self):
 
   @property
   def extra_jvm_options(self):
-    # TODO(jtrobec): implement code coverage using jacoco
-    return []
+    agent_option = '-javaagent:{agent}=destfile={destfile}'.format(agent=self._agent_path,
+                                                                   destfile=self._coverage_datafile)
+    return [agent_option]
+
+  @staticmethod
+  def is_coverage_target(tgt):
+    return (tgt.is_java or tgt.is_scala) and not tgt.is_test and not tgt.is_synthetic
 
   def report(self, execution_failed_exception=None):
-    # TODO(jtrobec): implement code coverage using jacoco
-    pass
+    if execution_failed_exception:
+      self._context.log.warn('Test failed: {0}'.format(execution_failed_exception))
+      if self._coverage_force:
+        self._context.log.warn('Generating report even though tests failed.')
```

Comment:
Message should probably indicate why ("because flag is set").

Reply:
ok, updated.